### DISCUSSION TITLE
[Idea] List of performance issues on PrestaShop software that are worth investigating

### DISCUSSION THREAD

**Discussion by matthieu-rolland** [PrestaShop/PrestaShop#35969] [Ideas]
### The problem

This is a copy/retranscription of a slack discussion about performance issues worth investigating in the PrestaShop software. 

- **translations page**: it seems it rescans translations every time you search for a new phrase, even if you made a typo, [see discussion here](https://github.com/PrestaShop/PrestaShop/discussions/35557)

- **stock manager performance with a lot of order**: #28116,  @kpodemski  tested [one of the community PRs](https://github.com/PrestaShop/PrestaShop/pull/34863) and it works correctly, what we’d need to do is making it cleaner. It drastically improves performance of creating an order (last step in PaymentModule::validateOrder) and managing the stock in BO ([issue](https://github.com/PrestaShop/PrestaShop/issues/14703)) We know mid-sized merchants use ERPs, but this also impacts perf of front office as mentioned

- **more products in shopping cart**: we know it’s a real pain for everyone working with PrestaShop, insightful discussion here: https://github.com/PrestaShop/PrestaShop/issues/14979 - watch out: this is not only about CartPresenter, this is about the Cart::getProducts method as well, and the problem we have by doing three-level querying of products sometimes, query from Cart::getProducts, then Product::getProductProperties, then we “present” the results, and we perform even more queries, one more product in cart is approx. 300-400 queries more sometimes - this is something what could make a great marketing material if fixed - one more hint: it is worth browsing phrase “slow cart” and “cart performance” even for closed issues because we mark them as duplicate and some of the helpful insights are almost lost ([example here](https://github.com/PrestaShop/PrestaShop/issues/24978))

- **autoupgrade module** scans entire shop to create xml file to determine what to upgrade, etc., this causes a few minutes loading time of upgrade screen (https://github.com/PrestaShop/PrestaShop/issues/30913 issue not reproduced but I’m sure it’s there because I’ve had it myself) - solution is to stop scanning some of the directories like /img/p

- **PrestaShopLogger, possible quick-win**: PrestaShopLogger can cause performance impacts in FO as many module developers use it incorrectly and log actions in front office, then if you have millions of records in logger it is causing visible performance problem if addLog method is called (more details and possible solutions: https://github.com/PrestaShop/PrestaShop/issues/27887)

- **Generation of pdf invoice**: when selected from date to date, it can be way too long when there are many invoices
 is having max execution time issues with 400 invoices.
Possible solution: generation per batch (let's say 10 by 10) in several ajax queries, with a progression bar in the UI, so that there's no max execution time issues and the user sees what's happening. @elyesnaouar also thinks that SQL queries can be improved on this.

### The solution

Investigate and fix if possible;

### Do you plan to work on this subject?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [ ] I'm willing to submit a Pull Request that implements this feature.
- [ ] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by matks** [comment]
https://github.com/PrestaShop/PrestaShop/issues/35978 just got posted 😄

**Comment by matks** [comment]
Some interesting links https://github.com/PrestaShop/PrestaShop/issues/19689 and https://github.com/PrestaShop/PrestaShop/issues/19701

**Comment by matks** [comment]
Check this out https://github.com/PrestaShop/PrestaShop/discussions/36004#discussioncomment-9228128

**Comment by kpodemski** [comment]
@matthieu-rolland @matks 

A few extra insights for you guys ;-)

1. Quick-win: Built-in notifications greatly slow down the back office loading time. Even if I disable them in the Performance -> Administration, there is still an AJAX request to /common/notifications. `CommonController` is not registered as a service, so I cannot override it. `notifications.ts` , and `AdminController::setMedia()` don't respect this setting. Removing notifications removes random slowness of the first load of the page.
2. ChoiceProviders for Orders are very slow (Tested on a store with 650k orders). I had to override `OrderCountriesChoiceProvider` because the query takes 700ms, and I couldn't manage to optimize the query inside. Same with `CarrierForOrderChoiceProvider`, it produces hundreds of extra queries on a single order (https://capture.dropbox.com/JmPSz9XtL70ahm7p). What's worse is that ChoiceProviders are injected into `OrderGridDefinitionFactory`, so I cannot easily remove the "Country name" column, I have to override the service.
3. `OrderQueryBuilder::getCountQueryBuilder` is very slow, around 0.5s. The workaround is not to use joins if not necessary.

**Reply by kpodemski** [reply]
after some tinkering with the code I managed to get down from 3s to 150-250ms for 650k orders listing page :)

**Reply by Hlavtox** [reply]
@kpodemski It would be awesome if you could contribute it to the core, so we have some useful changes there. :-)

**Reply by kpodemski** [reply]
@Hlavtox, some of those are just workarounds, not real solutions. Apart from notifications, the rest is probably not usable for global use, for example `OrderCountriesChoiceProvider`. If you have a lot of orders, you need to scan all of the address IDs for them to determine the country. A solution could be to record the delivery country name in the orders table. I will suggest some of the changes, but we know how it is and how the process looks with performance optimizations... :)

**Comment by Prestaworks** [comment]
When deleting products from a store with a large amount of products in the same category, the delete process is slow due to 
public static function cleanPositions($id_category, $position = 0)

where every other product than the one deleted in the same category gets an update on date_upd. This can cause thousands of update queryies, casuing the deletion of one product to take over two minutes.

`                $return &= Db::getInstance()->execute(
                    'UPDATE `' . _DB_PREFIX_ . 'product` p' . Shop::addSqlAssociation('product', 'p') . '
                    SET p.`date_upd` = "' . date('Y-m-d H:i:s') . '", product_shop.`date_upd` = "' . date('Y-m-d H:i:s') . '"
                    WHERE p.`id_product` = ' . (int) $result[$i]['id_product']
                );`

and

`
            for ($i = 0; $i < $total; ++$i) {
                $return &= Db::getInstance()->execute(
                    'UPDATE `' . _DB_PREFIX_ . 'product` p' . Shop::addSqlAssociation('product', 'p') . '
                    SET p.`date_upd` = "' . date('Y-m-d H:i:s') . '", product_shop.`date_upd` = "' . date('Y-m-d H:i:s') . '"
                    WHERE p.`id_product` = ' . (int) $result[$i]['id_product']
                );
            }`

these updates could probobly be removed

**Comment by elyesnaouar** [comment]
links to https://github.com/PrestaShop/PrestaShop/issues/39069
